Author Name  - Vijay Kumar Devalla

ASU ID - 1229148936

File creation date - Feb 16th, 2024

In [1]:
import pandas as pd

In [2]:


data = pd.read_csv('/Users/vijaykumardevalla/Documents/Unstructured Data Analytics/Assignment 5/restaurant_reviews_az.csv')

In [3]:
data['Sentiment'].unique()

array([1, 0])

In [4]:
data.head()

review_id                 user_id             business_id  \
0  IVS7do_HBzroiCiymNdxDg  fdFgZQQYQJeEAshH4lxSfQ  sGy67CpJctjeCWClWqonjA   
1  QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2  oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3  E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
4  Rd222CrrnXkXukR2iWj69g  LPxuausjvDN88uPr-Q4cQA  CA5BOxKRDPGJgdUQ8OUOpw   

   stars  useful  funny  cool  \
0      3       1      1     0   
1      5       1      1     1   
2      5       1      0     0   
3      5       0      0     0   
4      4       1      0     0   

                                                text             date  \
0  OK, the hype about having Hatch chili in your ...  1/27/2020 22:59   
1  Pandemic pit stop to have an ice cream.... onl...   4/19/2020 5:33   
2  I was lucky enough to go to the soft opening a...  2/29/2020 19:43   
3  I've gone to claim Jumpers all over the US and...  3/14/2020 21:47   
4  If you haven't been  to Maynard's kitchen, it'...  1/17/2020 20:32   

   Sentiment  
0          1  
1          1  
2          1  
3          1  
4          1

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48147 entries, 0 to 48146
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   review_id    48147 non-null  object
 1   user_id      48147 non-null  object
 2   business_id  48147 non-null  object
 3   stars        48147 non-null  int64 
 4   useful       48147 non-null  int64 
 5   funny        48147 non-null  int64 
 6   cool         48147 non-null  int64 
 7   text         48147 non-null  object
 8   date         48147 non-null  object
 9   Sentiment    48147 non-null  int64 
dtypes: int64(5), object(5)
memory usage: 3.7+ MB


In [6]:
df = data.head(1000)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   review_id    1000 non-null   object
 1   user_id      1000 non-null   object
 2   business_id  1000 non-null   object
 3   stars        1000 non-null   int64 
 4   useful       1000 non-null   int64 
 5   funny        1000 non-null   int64 
 6   cool         1000 non-null   int64 
 7   text         1000 non-null   object
 8   date         1000 non-null   object
 9   Sentiment    1000 non-null   int64 
dtypes: int64(5), object(5)
memory usage: 78.3+ KB


In [8]:
df.head()

review_id                 user_id             business_id  \
0  IVS7do_HBzroiCiymNdxDg  fdFgZQQYQJeEAshH4lxSfQ  sGy67CpJctjeCWClWqonjA   
1  QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2  oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3  E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
4  Rd222CrrnXkXukR2iWj69g  LPxuausjvDN88uPr-Q4cQA  CA5BOxKRDPGJgdUQ8OUOpw   

   stars  useful  funny  cool  \
0      3       1      1     0   
1      5       1      1     1   
2      5       1      0     0   
3      5       0      0     0   
4      4       1      0     0   

                                                text             date  \
0  OK, the hype about having Hatch chili in your ...  1/27/2020 22:59   
1  Pandemic pit stop to have an ice cream.... onl...   4/19/2020 5:33   
2  I was lucky enough to go to the soft opening a...  2/29/2020 19:43   
3  I've gone to claim Jumpers all over the US and...  3/14/2020 21:47   
4  If you haven't been  to Maynard's kitchen, it'...  1/17/2020 20:32   

   Sentiment  
0          1  
1          1  
2          1  
3          1  
4          1

In [9]:
df['Sentiment'].value_counts()

Sentiment
1    762
0    238
Name: count, dtype: int64

In [10]:
data['Sentiment'].value_counts()

Sentiment
1    33268
0    14879
Name: count, dtype: int64

In [11]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.callbacks import ModelCheckpoint
from tensorflow.keras.utils import plot_model

In [12]:
!pip install spacy


In [13]:
!python -m spacy download en_core_web_sm


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 18.7 MB/s eta 0:00:0000:0100:01
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [14]:
import spacy
nlp = spacy.load("en_core_web_sm")


In [15]:
# TF-IDF Vectorization
vectorizer = TfidfVectorizer(max_features=1000)
X_tfidf = vectorizer.fit_transform(data['text']).toarray()
y_tfidf = data['Sentiment'].values

In [16]:
# Define the ANN model
model = Sequential([
    Dense(2000, activation='relu', input_shape=(1000,)),
    Dense(1000, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the model
model.compile(optimizer='sgd', loss='binary_crossentropy', metrics=['accuracy'])

# Visualize the model
#plot_model(model, to_file='model.png', show_shapes=True, show_layer_names=True)


In [17]:
from tensorflow.keras.callbacks import ModelCheckpoint
from sklearn.model_selection import train_test_split

# Assuming X_tfidf and y_tfidf are defined earlier and model is already defined and compiled

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_tfidf, y_tfidf, test_size=0.25, random_state=42)

# Initialize ModelCheckpoint
checkpoint = ModelCheckpoint('best_model.hdf5', monitor='val_accuracy', save_best_only=True)

# Fit the model and save the history
history = model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=2, batch_size=8, callbacks=[checkpoint])

# Extract the history of validation accuracy
val_acc = history.history['val_accuracy']

# Find the maximum validation accuracy achieved
max_val_acc = max(val_acc)

print(f"Best Validation Accuracy: {max_val_acc*100:.2f}%")

Epoch 1/2
4514/4514 [==============================] - 23s 5ms/step - loss: 0.5261 - accuracy: 0.7330 - val_loss: 0.3346 - val_accuracy: 0.8688
Epoch 2/2
  38/4514 [..............................] - ETA: 18s - loss: 0.3649 - accuracy: 0.8717

/Users/vijaykumardevalla/anaconda3/lib/python3.11/site-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


4514/4514 [==============================] - 22s 5ms/step - loss: 0.2631 - accuracy: 0.8941 - val_loss: 0.2323 - val_accuracy: 0.9099
Best Validation Accuracy: 90.99%


In [18]:
# Define criteria for classification
positive_threshold = 0.7  # Probabilities above this are considered positive
negative_threshold = 0.3  # Probabilities below this are considered negative
# Probabilities between negative_threshold and positive_threshold are considered neutral

# Print criteria for classification
print(f"Classification Criteria:")
print(f"Positive: Probability > {positive_threshold}")
print(f"Neutral: Probability between {negative_threshold} and {positive_threshold}")
print(f"Negative: Probability < {negative_threshold}\n")

# Define the list of review texts
reviews = [
    "The service is good, but location is hard to find. Sanitation is not very good with old facilities. Food served tasted extremely fishy, making us difficult to even finish it.",
    "The restaurant is definitely one of my favorites and of my family as well. I was especially impressed with my visit a few days ago. The place is clean, and you just need to wait for fewer than 10 minutes to get food served. And of course, the food is absolutely delicious!",
    "I appreciated the friendly staff. The food was good, not amazing. The service was not prompt but almost acceptable. A reliable spot for a regular meal, but nothing extraordinary."
]

# Vectorize the reviews
X_new = vectorizer.transform(reviews).toarray()

# Predict sentiments
predictions = model.predict(X_new)

# Print classification results and reasoning
for i, review in enumerate(reviews):
    sentiment = 'Neutral'  # Default to neutral
    if predictions[i] > positive_threshold:
        sentiment = 'Positive'
    elif predictions[i] < negative_threshold:
        sentiment = 'Negative'
    print(f"Review {i+1}: Sentiment {sentiment} with probability {predictions[i][0]}")


Classification Criteria:
Positive: Probability > 0.7
Neutral: Probability between 0.3 and 0.7
Negative: Probability < 0.3

1/1 [==============================] - 0s 40ms/step
Review 1: Sentiment Negative with probability 0.18975599110126495
Review 2: Sentiment Positive with probability 0.9916536808013916
Review 3: Sentiment Neutral with probability 0.5723163485527039


In [19]:
import pandas as pd
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Step 1: Reload Data

reviews = df['text'].values
y = df['Sentiment'].values  # Adjust column names as per your dataset

# Step 2: Tokenize and Pad Sequences
tokenizer = Tokenizer()
tokenizer.fit_on_texts(reviews)
sequences = tokenizer.texts_to_sequences(reviews)
X_padded = pad_sequences(sequences, maxlen=50)

# Preparing for Step 3: Load GloVe Embeddings (assuming you've downloaded 'glove.6B.zip' and extracted it)
embed_file = '/Users/vijaykumardevalla/Documents/Unstructured Data Analytics/Assignment 5/glove.6B/glove.6B.50d.txt'  # Make sure this path is correct and file is extracted

# Load GloVe 300d embeddings
embeddings_index = {}
with open(embed_file, 'r', encoding='utf-8') as f:
    for line in f:
        values = line.split()
        word = values[0]
        coefs = np.asarray(values[1:], dtype='float32')
        embeddings_index[word] = coefs

# Step 4: Create Embedding Matrix
vocab_size = len(tokenizer.word_index) + 1  # Plus one for padding token
embedding_dim = 50
embedding_matrix = np.zeros((vocab_size, embedding_dim))

for word, i in tokenizer.word_index.items():
    embedding_vector = embeddings_index.get(word)
    if embedding_vector is not None:
        embedding_matrix[i] = embedding_vector

# Verification
print("Shape of embedding matrix:", embedding_matrix.shape)  # Should be (vocab_size, 300)

# At this point, you have:
# - X_padded: Tokenized and padded sequences ready for model input
# - embedding_matrix: Prepared embedding matrix using GloVe 300d embeddings
# - y: Target sentiment labels


Shape of embedding matrix: (6176, 50)


In [20]:
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping


In [21]:
from tensorflow.keras.mixed_precision import set_global_policy

# Enable mixed precision
set_global_policy('mixed_float16')





The dtype policy mixed_float16 may run slowly because this machine does not have a GPU. Only Nvidia GPUs with compute capability of at least 7.0 run quickly with mixed_float16.
If you will use compatible GPU(s) not attached to this host, e.g. by running a multi-worker model, you can ignore this warning. This message will only be logged once


In [22]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Flatten, Dense
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping
from tensorflow.keras.optimizers import Adam

# Assuming X_train, X_test, y_train, y_test are already defined from the previous cell

# Enable mixed precision
from tensorflow.keras.mixed_precision import set_global_policy
set_global_policy('mixed_float16')

model = Sequential([
    Embedding(input_dim=vocab_size, output_dim=50, input_length=1000, weights=[embedding_matrix], trainable=False),
    Flatten(),
    Dense(1024, activation='relu'),  # Reduced complexity
    Dense(512, activation='relu'),  # Reduced complexity
    Dense(1, activation='sigmoid')  # Output layer for binary classification
])

# Compile the model with Adam optimizer for better performance
model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

# Prepare callbacks for saving the best model and early stopping
checkpoint = ModelCheckpoint('best_ANN_model_optimized.hdf5', monitor='val_accuracy', save_best_only=True)
early_stopping = EarlyStopping(monitor='val_loss', patience=3, mode='min', restore_best_weights=True)

# Fit the model
history = model.fit(
    X_train, y_train,
    validation_data=(X_test, y_test),
    epochs=2,  # As specified
    batch_size=8,  # As specified
    callbacks=[checkpoint, early_stopping]
)

# Optional: Print the best validation accuracy
val_acc = max(history.history['val_accuracy'])
print(f"Best Validation Accuracy: {val_acc*100:.2f}%")


Epoch 1/2
4514/4514 [==============================] - 733s 162ms/step - loss: 0.6255 - accuracy: 0.6916 - val_loss: 0.6202 - val_accuracy: 0.6888
Epoch 2/2
4514/4514 [==============================] - 946s 210ms/step - loss: 0.6177 - accuracy: 0.6917 - val_loss: 0.6205 - val_accuracy: 0.6888
Best Validation Accuracy: 68.88%


The analysis showcases a pronounced disparity in the efficacy between TF-IDF and ANN models, with TF-IDF achieving a remarkable accuracy of 90.99%, significantly surpassing the ANN model's 68.88%. This performance differential underscores the effectiveness and efficiency of TF-IDF in specific analytical scenarios, particularly in tasks that benefit from a keyword-focused approach and do not necessitate extensive computational resources. Such findings emphasize the criticality of selecting an appropriate modeling approach tailored to the task's unique demands, computational availability, and dataset characteristics.






# RNN Model

In [24]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, Dense, Dropout
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping
from sklearn.model_selection import train_test_split

# Assuming X_padded, y, and embedding_matrix are prepared as per the previous instructions

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_padded, y, test_size=0.25, random_state=42)

# Define the RNN model
model = Sequential([
    Embedding(input_dim=vocab_size, output_dim=50, weights=[embedding_matrix], input_length=50, trainable=False),
    SimpleRNN(2000, return_sequences=True, activation='relu'),  # First RNN layer with return_sequences=True
    SimpleRNN(1000, activation='relu'),  # Second RNN layer
    Dense(1, activation='sigmoid')  # Output layer for binary classification
])

# Compile the model
model.compile(optimizer='sgd', loss='binary_crossentropy', metrics=['accuracy'])

# Visualize the model architecture
#tf.keras.utils.plot_model(model, to_file='RNN_model_architecture.png', show_shapes=True, show_layer_names=True)

# Prepare callbacks for saving the best model based on validation accuracy and for early stopping
checkpoint = ModelCheckpoint('best_RNN_model.hdf5', monitor='val_accuracy', save_best_only=True, mode='max')
early_stopping = EarlyStopping(monitor='val_loss', patience=3, mode='min', restore_best_weights=True)

# Fit the model
history = model.fit(X_train, y_train, validation_data=(X_test, y_test),
                    epochs=2, batch_size=8, callbacks=[checkpoint, early_stopping])

# Print the best validation accuracy
val_acc = max(history.history['val_accuracy'])
print(f"Best Validation Accuracy: {val_acc*100:.2f}%")


Epoch 1/2
94/94 [==============================] - 31s 328ms/step - loss: 0.5566 - accuracy: 0.7613 - val_loss: 0.5581 - val_accuracy: 0.7480
Epoch 2/2


/Users/vijaykumardevalla/anaconda3/lib/python3.11/site-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


94/94 [==============================] - 31s 333ms/step - loss: 0.5177 - accuracy: 0.7667 - val_loss: 0.5491 - val_accuracy: 0.7480
Best Validation Accuracy: 74.80%


# LSTM

In [26]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping
from sklearn.model_selection import train_test_split

# Assuming X_padded, y, and embedding_matrix are prepared as per the previous instructions

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_padded, y, test_size=0.25, random_state=42)

# Define the LSTM model
model = Sequential([
    Embedding(input_dim=vocab_size, output_dim=50, weights=[embedding_matrix], input_length=50, trainable=False),
    LSTM(2000, return_sequences=True, activation='relu'),  # First LSTM layer with return_sequences=True
    LSTM(1000, activation='relu'),  # Second LSTM layer
    Dense(1, activation='sigmoid')  # Output layer for binary classification
])

# Compile the model
model.compile(optimizer='sgd', loss='binary_crossentropy', metrics=['accuracy'])

# Visualize the model architecture
#tf.keras.utils.plot_model(model, to_file='LSTM_model_architecture.png', show_shapes=True, show_layer_names=True)

# Prepare callbacks for saving the best model based on validation accuracy and for early stopping
checkpoint = ModelCheckpoint('best_LSTM_model.hdf5', monitor='val_accuracy', save_best_only=True, mode='max')
early_stopping = EarlyStopping(monitor='val_loss', patience=3, mode='min', restore_best_weights=True)

# Fit the model
history = model.fit(X_train, y_train, validation_data=(X_test, y_test),
                    epochs=2, batch_size=8, callbacks=[checkpoint, early_stopping])

# Print the best validation accuracy
val_acc = max(history.history['val_accuracy'])
print(f"Best Validation Accuracy: {val_acc*100:.2f}%")


Epoch 1/2
94/94 [==============================] - 159s 2s/step - loss: 0.5766 - accuracy: 0.7667 - val_loss: 0.5647 - val_accuracy: 0.7480
Epoch 2/2


/Users/vijaykumardevalla/anaconda3/lib/python3.11/site-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


94/94 [==============================] - 159s 2s/step - loss: 0.5469 - accuracy: 0.7667 - val_loss: 0.5632 - val_accuracy: 0.7480
Best Validation Accuracy: 74.80%


# Compare and comment on your observations on the performance of different deep learning models (i.e., ANN vs. RNN vs. LSTM).

Results indicate that RNN and LSTM models, both achieving a best validation accuracy of 74.80%, outperform the ANN model with word embeddings, which reached 68.88% accuracy. This suggests that the sequential processing capabilities of RNNs and LSTMs, which allow them to better understand the context and dependencies within the text data, provide a clear advantage for tasks requiring nuanced understanding of language, such as sentiment analysis. Despite the simplicity and lower computational demands of the ANN, its performance lag suggests that capturing temporal dynamics in text data is crucial for higher accuracy in this context.

I took help from chatgpt